draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,175 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+    except:
+        return Rejected()
+~~~
+
+If the state is `Rejected`, then processing halts. Otherwise, if the state is
+`Continued`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), The Leader sends its prep share
+  to the Helper, who computes the prep message locally (the computations are
+  specified in the following sections), computes its output share, then sends
+  the prep message to the Leader. Preparation requires just one request from
+  the Leader to the Helper.
+
+* For a 2-round VDAF (e.g., Poplar1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   state: State,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   inbound: Optional[Message],
+                   ) -> (State, Optional[Message]):
+    if type(inbound) == InitializeMessage:
+        prep_msg, peer_prep_share = None, inbound.prep_share
+    elif type(inbound) == ContinueMessage:
+        prep_msg, prep_share = inbound.prep_msg, inbound.prep_share
+    elif type(inbound) == FinishMessage:
+        prep_msg, peer_prep_share = inbound.prep_msg, None
+    else:
+        prep_msg, peer_prep_share = None, None
+
+    try:
+        # If prep_msg == None then this is the start of the
+        # first round. Otherwise, this is the prep message
+        # from the end of the previous round.
+        out = Vdaf.prep_next(host_prep_state, prep_msg)
+        if type(out) == Vdaf.OutShare:
+            # The Leader reaches this state upon processing a
+            # response from the Helper, so there is no outbound
+            # message.
+            outbound = None if is_leader else FinishedMessage()
```

Comment:
We should probably enumerate and describe the different message types before this pseudocode, because the difference between FinishMessage and FinishedMessage is easy to miss.

Reply:
I agree -- it would be very nice to describe these messages along with their purpose/contents.

Reply:
Done! While at it I realized we maybe don't need `FinishedMessage` at this level of abstraction-- We could just have the Helper respond with `None`. Of course, in DAP we'll have to distinguish between "finish(prep_msg)" and "finished" in the wire protocol.

Let me know what you think -- if desired I can add this back in, I don't feel strongly about it either way.